metron-platform/metron-writer/src/main/java/org/apache/metron/writer/bolt/BulkMessageWriterBolt.java

```diff
@@ -92,22 +177,51 @@ public void prepare(Map stormConf, TopologyContext context, OutputCollector coll
       configurationTransformation = x -> x;
     }
     try {
-      bulkMessageWriter.init(stormConf
-                            , configurationTransformation.apply(new IndexingWriterConfiguration(bulkMessageWriter.getName(), getConfigurations()))
-                            );
+      WriterConfiguration writerconf = configurationTransformation.apply(
+              new IndexingWriterConfiguration(bulkMessageWriter.getName(), getConfigurations()));
+      if (defaultBatchTimeout == 0) {
+        //This means getComponentConfiguration was never called to initialize defaultBatchTimeout,
+        //probably because we are in a unit test scenario.  So calculate it here.
+        BatchTimeoutHelper timeoutHelper = new BatchTimeoutHelper(writerconf::getAllConfiguredTimeouts, batchTimeoutDivisor);
+        defaultBatchTimeout = timeoutHelper.getDefaultBatchTimeout();
+      }
+      writerComponent.setDefaultBatchTimeout(defaultBatchTimeout);
+      bulkMessageWriter.init(stormConf, writerconf);
     } catch (Exception e) {
       throw new RuntimeException(e);
     }
   }
 
+  /**
+   * Used only for unit testing.
+   */
+  public void prepare(Map stormConf, TopologyContext context, OutputCollector collector, Clock clock) {
+    prepare(stormConf, context, collector);
+    writerComponent.withClock(clock);
+  }
+
   @SuppressWarnings("unchecked")
   @Override
   public void execute(Tuple tuple) {
-    JSONObject message = (JSONObject) messageGetStrategy.get(tuple);
-    String sensorType = MessageUtils.getSensorType(message);
     try
     {
-      WriterConfiguration writerConfiguration = configurationTransformation.apply(new IndexingWriterConfiguration(bulkMessageWriter.getName(), getConfigurations()));
+      if (isTick(tuple)) {
+        if (!(bulkMessageWriter instanceof WriterToBulkWriter)) {
+          //WriterToBulkWriter doesn't allow batching, so no need to flush on Tick.
+          LOG.debug("Flushing message queues older than their batchTimeouts");
+          writerComponent.flushTimeouts(bulkMessageWriter, configurationTransformation.apply(
+                  new IndexingWriterConfiguration(bulkMessageWriter.getName(), getConfigurations()))
+                  , messageGetStrategy);
+        }
+        collector.ack(tuple);
```

Comment:
Good idea, will change to separate "try" wrappers around the tick tuple vs message tuple processing.